drivers/greth.c

```diff
@@ -0,0 +1,631 @@
+
+	state->mmio->CTRL |= GRETH_CTRL_RE;
+}
+
+
+static int greth_createResources(greth_state_t *state)
+{
+	state->irq_lock = (handle_t)-1;
+	state->irq_cond = (handle_t)-1;
+
+	if (mutexCreate(&state->irq_lock) < 0) {
+		return -1;
+	}
+	if (condCreate(&state->irq_cond) < 0) {
+		return -1;
+	}
+	return 0;
+}
+
+
+static void greth_cleanupResources(greth_state_t *state)
+{
+	if (state->irq_lock != (handle_t)-1) {
+		resourceDestroy(state->irq_lock);
+	}
+	if (state->irq_cond != (handle_t)-1) {
+		resourceDestroy(state->irq_cond);
+	}
+}
+
+
+static void greth_reset(greth_state_t *state)
+{
+	state->mmio->CTRL = GRETH_CTRL_RS;
+	do {
+	} while ((state->mmio->CTRL & GRETH_CTRL_RS) != 0);
+}
+
+
+static int greth_initDevice(greth_state_t *state, int irq, bool mdio)
+{
+	state->mmio = physmmap(state->dev_phys_addr, _PAGE_SIZE);
+	if (state->mmio == MAP_FAILED) {
+		return -ENOMEM;
+	};
+
+	greth_reset(state);
+
+#if (GRETH_EDCL != 1)
+	greth_disableEdcl(state);
+#endif
+
+	int err = greth_initRings(state);
+	if (err != 0) {
+		return err;
+	}
+	greth_debug_printf(state, "Initialized GRETH Rings");
+
+
+	if (mdio) {
+		err = register_mdio_bus(&greth_mdio_ops, state);
+		if (err < 0) {
+			greth_printf(state, "Can't register MDIO bus:  %s (%d)", strerror(-err), err);
+			return err;
+		}
+
+		greth_debug_printf(state, "MDIO bus %d", err);
+	}
+
+	err = greth_createResources(state);
+	if (err != 0) {
+		greth_printf(state, "Couldn't create resources: %s (%d)", strerror(-err), err);
+		greth_cleanupResources(state);
+		return err;
+	}
+
+	err = interrupt(irq, greth_irqHandler, state, state->irq_cond, NULL);
+	if (err != 0) {
+		greth_printf(state, "Couldn't register interrupt handler: %s (%d)", strerror(-err), err);
+		greth_cleanupResources(state);
+		return err;
+	}
+	greth_debug_printf(state, "Interrupt handler initialized successfully");
+
+	err = beginthread(greth_irqThread, 1, state->irq_stack, sizeof(state->irq_stack), state);
+	if (err != 0) {
+		greth_printf(state, "Couldn't begin interrupt thread: %s (%d)", strerror(-err), err);
+		greth_cleanupResources(state);
+		return err;
+	}
+
+	greth_setCardMac(state);
+	greth_readCardMac(state);
+	greth_showCardId(state);
+
+	net_refillRx(&state->rx, ETH_PAD_SIZE);
+	greth_startGreth(state);
+
+	return err;
+};
+
+
+static int greth_netifInit(struct netif *netif, char *cfg)
+{
+	greth_state_t *state;
+	char *p;
+	int err, irq;
+	bool mdio = true;
+
+	netif->linkoutput = greth_netifOutput;
+	state = netif->state;
+	state->netif = netif;
+
+	if (cfg == NULL) {
+		return -EINVAL;
+	}
+
+	/* base addr */
+	state->dev_phys_addr = strtoul(cfg, &p, 0);
+	if (*cfg == '\0' || *p++ != ':') {
+		return -EINVAL;
+	}
+
+	/* irq */
+	cfg = p;
+	irq = strtoul(cfg, &p, 0);
+	if (*cfg == '\0' || (*p != '\0' && *p++ != ':') || irq < 0) {
+		return -EINVAL;
+	}
+
+	/* MDIO and PHY opts */
+	cfg = NULL;
+	while (p != NULL && *p != '\0') {
+		cfg = strchr(p, ':');
+		if (cfg != NULL) {
+			*cfg++ = '\0';
```

Comment:
MISRA: statements should not have side effects
```suggestion
			*cfg = '\0';
			cfg++;
```